### ISSUE TITLE
Callbacks are running on the main thread, blocking the UI

### ISSUE THREAD

**Issue by EduardMe** [dropbox/SwiftyDropbox#255]
Hey there,

I have implemented Dropbox Sync into my document based app ([NotePlan](https://noteplan.co)) as beta currently. The sync seems to work, but there is one big problem. All the callbacks seem to run on the main thread and thus briefly blocking the user interface. Files are constantly being synced up, so a list of files needs to be loaded frequently and compared with the local set of files (we maintain a set of local files where changes by the user are saved, then we scan them for updates and upload them to the user's dropbox. Equally do we check for updates on the dropbox account).

Here is a snippet:

```
DropboxClientsManager.authorizedClient?.files.download(path: source.relativePath)
            .response(queue: DispatchQueue.global(), completionHandler: { (file, error) in
                if let file = file {
                    do { try file.1.write(to: destination)
                    } catch { print("could not write data from dropbox file into the local folder \(error)") }
                    
                    // Set the modification date, so its the same as online
                    do {
                        try FileManager.default.setAttributes([FileAttributeKey.modificationDate: file.0.clientModified], ofItemAtPath: destination.relativePath)
                    } catch { }
                    
                    self.downloadingFiles.removeAll() { $0 == destination.lastPathComponent }
                    
                    // Update the cache
                    self.localCache[destination] = file.0.clientModified
                    completed?()
                }
            })
```

And everything is wrapped in:

```
DispatchQueue.global().async {
    // The dropbox sync logic
}
```

Am I doing something wrong or is there a way to change the thread? Maybe `DispatchQueue.global()` is wrong? Not sure.

Let me know if you need more details! Thanks!

**Comment by EduardMe** [comment]
Thank yo ufos the quick reply. I can confirm, when I run `Thread.isMainThread` during the callbacks, it returns always false. Still something is blocking the UI. I'm investigating it further and closing this issue if I don't find anything today.

**Comment by EduardMe** [comment]
I think I got something. The UI becomes unresponsive when I upload files in bulk:

```
DropboxClientsManager.authorizedClient?.files
            .batchUploadFiles(fileUrlsToCommitInfo: info, queue: DispatchQueue.global(), responseBlock: { (results, error, errors) in
                print("actual batch uploading, is main thread: \(Thread.isMainThread)")
                if let error = error {
                    print("Error: \(error)")
                }
                
                self.uploadingFiles.removeAll()
                self.uploadingCache[base]?.removeAll()
            })
```

 Digging further, I found this in SwiftyDropbox -> CustomRoutes.swift:

```
func batchFinishUponCompletion(uploadData: BatchUploadData) {
        // wait for all upload calls to complete and then batch "finish" all uploaded files
        // with one call to `upload_session/finish_batch`
        uploadData.uploadGroup.notify(queue: DispatchQueue.main) {
            uploadData.finishArgs.sort { $0.commit.path < $1.commit.path }
            
            self.uploadSessionFinishBatch(entries: uploadData.finishArgs).response { result, error in
                if let result = result {
                    switch result {
                    case .asyncJobId(let asyncJobId):
                        sleep(1)
                        self.queryJobStatus(uploadData: uploadData, asyncJobId: asyncJobId, retryCount: 2)
                    default:
                        break;
                    }
                } else {
                    uploadData.queue.async {
//                        uploadData.responseBlock(nil, nil, error, uploadData.fileUrlsToRequestErrors)
                    }
                }
            }
        }
    }
```

Note that `uploadData.uploadGroup.notify(queue: DispatchQueue.main)` uses the main thread.
And the response block of `self.uploadSessionFinishBatch` is not defining anything for the `queue` attribute, making it main thread. There are more internal functions which don't use the `queue` attribute. I'm trying to set it in the Dropbox code and check it again. But there I got `true` when calling `print Thread.isMainThread` during execution. This is only when uploading files so far.

**Comment by EduardMe** [comment]
Yes, it seems I got it fixed this way. Basically I checked for all `.response` and `DispatchQueue.main` occurrences and fixed it wherever possible with the `uploadData.queue` parameter. Unresponsiveness is gone so far. Shall I send the file or upload the code somewhere?

**Comment by greg-db** [comment]
Oh, I see, thanks! Yes, `batchUploadFiles` is a specially-written helper and works differently than the standard methods (like `download`). It looks like it is using the main thread. I'll bring this up with the team and ask them to change that.

We do welcome PRs if you want to share your fix. Thanks again.

**Comment by greg-db** [comment]
This should be fixed in v8.3.0.
